activesupport/lib/active_support/cache/coder.rb

```diff
@@ -0,0 +1,142 @@
+# frozen_string_literal: true
+
+require_relative "entry"
+
+module ActiveSupport
+  module Cache
+    class Coder # :nodoc:
+      def initialize(serializer, compressor, legacy_serializer: false)
+        @serializer = serializer
+        @compressor = compressor
+        @legacy_serializer = legacy_serializer
+      end
+
+      def dump(entry)
+        return @serializer.dump(entry) if @legacy_serializer
+
+        dump_compressed(entry, Float::INFINITY)
+      end
+
+      def dump_compressed(entry, threshold)
+        return @serializer.dump_compressed(entry, threshold) if @legacy_serializer
+
+        # If value is a string with a supported encoding, use it as the payload
+        # instead of passing it through the serializer.
+        if type = type_for_string(entry.value)
+          payload = binary_compatible_string(entry.value)
+        else
+          type = OBJECT_DUMP_TYPE
+          payload = serialize(entry.value)
+        end
+
+        # Compress payload if it meets the size threshold. (Note that `type < 0`
+        # signifies a compressed payload with `type.abs` as its actual type.)
+        if compressed = try_compress(payload, threshold)
+          payload = compressed
+          type = -type
```

Comment:
So using the sign bit to mark compressed it clever, but I think I'd prefer a more explicit bitmap:

```ruby
COMPRESSED_FLAG = 0x80

type = type | COMPRESSED_FLAG
```

This mean the type template would become `C` (unsigned byte) not `c` (signed byte)